XMonad/WorkspaceLayout/Grid.hs

```diff
@@ -0,0 +1,326 @@
+
+
+-- | Glue together a column of workspaces
+column :: forall ftd. IsFormatted ftd => Dims -> Int -> String -> Mapping ftd
+column (Dims { height }) x name =
+  Mapping . fold $ do
+    y <- [0 .. height - 1]
+    let topLeft = XY x 0
+    pure $ Map.singleton (XY x y) (doFormat @ftd topLeft name)
+
+
+
+
+data Coord = XY { x :: Int, y :: Int }
+  deriving (Show, Ord, Eq, Generic)
+
+data Wrapping = Wrapping
+  { wrapX :: Bool
+  , wrapY :: Bool
+  }
+  deriving (Show, Generic)
+
+data State = State
+  { mapping  :: SomeMapping
+      -- ^ Coordinate -> WorkspaceId mapping
+  , labelf   :: Coord -> Maybe String
+      -- ^ Labels
+  , wrapping :: Wrapping
+      -- ^ Wrapping mode
+  , coord    :: Coord
+      -- ^ Current coordinate
+  } deriving (Generic)
+
+instance St.OneState State where
+  type Mod State = State -> State
+  merge ma s = pure (ma s)
+  defaultState = State
+    { mapping = SomeMapping (grid $ Dims 5 5)
+    , wrapping = Wrapping False False
+    , coord = XY 0 0
+    , labelf = const Nothing
+    }
+
+-- Wrap a coordinate around the x/y axes according to the configured wrapping mode
+wrap :: State -> (Coord -> Coord)
+wrap state =
+
+  appEndo . fromMaybe (Endo id) $ do
+    xRange <- range x (mapping state)
+    yRange <- range y (mapping state)
+
+    pure $ guard (state & wrapping & wrapX) (Endo $ \c -> c { x = affineMod xRange (x c) })
+        <> guard (state & wrapping & wrapY) (Endo $ \c -> c { y = affineMod yRange (y c) })
+
+  where
+
+  guard :: Monoid m => Bool -> m -> m
+  guard = \case { True -> id; False -> const mempty }
+
+
+
+-- |
+--
+-- The call @move f@ replaces the current coordinate to @f currentCoord@
+--
+-- If @f currentCoord@ is out-of-bounds, do nothing
+--
+-- Use this to move around workspaces
+move :: (Coord -> Coord) -> X ()
+move = update $ \coord wid -> do
+  St.modify $ \st -> st { coord = coord }
+  windows (greedyView wid)
+
+
+-- |
+--
+-- The call @swap f@ moves the selected window to @f currentCoord@
+--
+-- If @f currentCoord@ is out-of-bounds, do nothing
+swap :: (Coord -> Coord) -> X ()
+swap = update $ \_ wid -> windows (shift wid)
+
+update :: (Coord -> WorkspaceId -> X ()) -> (Coord -> Coord) -> X ()
+update act f = do
+  state@State { coord, mapping } <- St.get
+  let coord' = coord & f & wrap state
+  case Map.lookup coord' (getTheMap mapping) of
+    Nothing -> pure ()
+    Just wid -> do
+     act coord' wid
+
+
+data Init = Init
+  { initMapping  :: SomeMapping
+      -- ^ Physical layout
+  , initWrapping :: Wrapping
+      -- ^ Wrapping mode
+  , initLabelf   :: Coord -> Maybe String
+      -- ^
+      -- Displayed to the left of the axis workspaces
+      -- When Nothing is returned, uses a fallback
+  }
+
+-- | Hook the grid layout into XMonad
+hook :: Init -> XConfig l -> XConfig l
+hook Init { initMapping, initWrapping, initLabelf } =
+  St.once @State
+    (\xc -> xc { XMonad.workspaces = workspaces })
```

Comment:
I'm torn about this interface. On the one hand, it's cool because you have a single entry point to hook the module to your configuration, which makes it easy to add and remove. On the other hand, this behavior might be confusing since it just throws away the workspaces in the initial config. It might be okay though since I can't think of an ergonomic behavior

Reply:
Yeah, it's kind of a fundamental limitation of my design that the code needs to have control over your `workspaces`.

The good news is that by initializing with a `Mapping 'Unformatted`, the user can choose what the workspace names should be. So if they already have a list of workspaces, with the right choice of `Init` value `hook` will not change those workspaces.

That fact could use some documentation though : P

Reply:
Also, it's worth noting that this "single entrypoint" is not as clean as it may seem. For effective use of `XMonad.WorkspaceLayouts.Grid` one has to (a) use `hook`; (b) modify their pretty-printer; and (c) setup keybindings

Reply:
IMO it's fine to replace workspaces. Might be nice to mention it in the docs, but it's okay to do it.

Reply:
`X.L.IndependentScreens` also rewrites `workspaces`, although it does so by having you call a function explicitly. Maybe there's something here in terms of design that could be used to simplify use of both? (`IndependentScreens` is currently pretty invasive and difficult to set up.)